### DISCUSSION TITLE
Multiple table names in Queryable and QueryableByName

### DISCUSSION THREAD

**Discussion by snf** [diesel-rs/diesel#3779] [Ideas]
In the issue https://github.com/diesel-rs/diesel/issues/1314 it was discussed to add multiple table names for `Insertable` which got implemented in https://github.com/diesel-rs/diesel/pull/3535

However, I'm having an issue with this because when I want to derive `Insertable` and `Queryable`/`QueryableByName` for a struct I need to use two different structs. I'm hacking around this implementing `From` one to the other to avoid too much code.

In the discussion I couldn't find any mention of why `Queryable` and `QueryableByName` were excluded.

- Queryable: it doesn't use the table name so the restriction could be lifted: https://github.com/diesel-rs/diesel/blob/master/diesel_derives/src/queryable.rs

- QueryableByName uses the table name to guess the types of the SQL columns but we could use the non-guessing path if there are multiple table names or compare that the type is the same in all the tables: https://github.com/diesel-rs/diesel/blob/master/diesel_derives/src/queryable_by_name.rs#L117 . Or other ideas?

There is also the discussion if we really want this as it may break backwards compatiblity once implemented if there are other features planned for the derives.

I can make a PR for either of those. I think implementing it for Queryable is trivial and it would just need a test.

**Comment by weiznich** [comment]
I think I would be fine with allowing that for `#[derive(Queryable)]` as `table_name` is not used there anyway.
For `#[derive(QueryableByName)]` we need to be a bit more careful as it might become something that's otherwise confusing for users, if it only "sometimes" works. I think we can workaround that by requiring that all fields have a `#[diesel(sql_type = …)]` attribute and otherwise generate an explicit error message explaining the problem.

**Comment by snf** [comment]
Added for Queryable in https://github.com/diesel-rs/diesel/pull/3782

Should I leave this open for QueryableByName or close and re-open in case I get to work on that one?

**Reply by weiznich** [reply]
Its fine to just leave it open
